rosbag2_transport/src/rosbag2_transport/rosbag2_transport.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rosbag2_transport/rosbag2_transport.hpp"
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <queue>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcl/graph.h"
+#include "rclcpp/rclcpp.hpp"
+#include "rcutils/time.h"
+
+#include "rosbag2_transport/logging.hpp"
+#include "rosbag2/sequential_reader.hpp"
+#include "rosbag2/types.hpp"
+#include "rosbag2/typesupport_helpers.hpp"
+#include "rosbag2/writer.hpp"
+
+#include "generic_subscription.hpp"
+#include "player.hpp"
+#include "replayable_message.hpp"
+#include "rosbag2_node.hpp"
+
+namespace rosbag2_transport
+{
+
+Rosbag2Transport::Rosbag2Transport()
+: reader_(std::make_shared<rosbag2::SequentialReader>()),
+  writer_(std::make_shared<rosbag2::Writer>())
+{}
+
+Rosbag2Transport::Rosbag2Transport(
+  std::shared_ptr<rosbag2::SequentialReader> reader, std::shared_ptr<rosbag2::Writer> writer)
+: reader_(std::move(reader)), writer_(std::move(writer)) {}
+
+void Rosbag2Transport::init()
+{
+  rclcpp::init(0, nullptr);
+}
+
+void Rosbag2Transport::shutdown()
+{
+  rclcpp::shutdown();
+}
+
+void Rosbag2Transport::record(
+  const StorageOptions & storage_options, const RecordOptions & record_options)
+{
+  writer_->open(storage_options);
+
+  auto rosbag2_transport = std::make_shared<Rosbag2Node>("rosbag2");
+
+  auto topics_and_types = record_options.all ?
+    rosbag2_transport->get_all_topics_with_types() :
+    rosbag2_transport->get_topics_with_types(record_options.topics);
+
+  if (topics_and_types.empty()) {
+    throw std::runtime_error("No topics found. Abort");
+  }
+
+  for (const auto & topic_and_type : topics_and_types) {
+    auto topic_name = topic_and_type.first;
+    auto topic_type = topic_and_type.second;
+
+    auto subscription = create_subscription(rosbag2_transport, topic_name, topic_type);
+    if (subscription) {
+      subscriptions_.push_back(subscription);
+      writer_->create_topic({topic_name, topic_type});
+    }
+  }
+
+  if (subscriptions_.empty()) {
+    throw std::runtime_error("No topics could be subscribed. Abort");
+  }
+
+  ROSBAG2_TRANSPORT_LOG_INFO("Waiting for messages...");
```

Comment:
I think this message can be deleted. It will always be printed, whether topics are available or actually have to be waited on.
I'd prefer an output to which topics I subscribed though. That's more helpful

Reply:
I agree, but it's not completely trivial (I have to store a vector of successfully subscribed topics and then concatenate them here to write them to the logging macro). I'd like to open an issue and do it later if that's okay with you.